Firmware/MVP/battsci.cpp

```diff
@@ -86,6 +98,169 @@ uint8_t BATTSCI_calculateChecksum( uint8_t frameSum )
+  //  20% = 0x11 0x48 -- Regen allowed, No Assist, Regen running even during idle
+  //  20% = 0x21 0x48 -- THIS IS ALSO 20%
+
+  // MCM can read and accept values outside this window, like 19% SoC or 85% SoC, but we don't need to use those.
+
+  /**
+    BATTSCI_evaluateSoCBytes evaluates what bytes LiBCM will send to the MCM to set the SoC, depending on the SoC we gave it
+
+    @param      evalSoC     Integer value of SoC in 0.1% increments beginning at 20%. 0 = 20.0%
+                            This should be between 0 and 600.
+                            Out of bounds values will still work, but may cause unexpected results.
+    @return                 This function does not return anything.  It modifies SoC_Bytes[] in place.
+  */
+
+  SoC_Bytes[0] = 0x00;
+  evalSoC += 0x48;			// MCM 2nd-byte SoC is 0x48 when SoC is 20%.  20% is our reference, so we need to add this first.
+
+  do {
+    SoC_Bytes[0] += 0x01;
+    evalSoC -= 128;			// MCM 2nd-byte SoC CANNOT exceed 0x7F so we need to subtract 128 from it over and over until 2nd-byte is < 0x80
+  }
+  while (evalSoC > 128);
+
+  SoC_Bytes[0] += 0x21;		// MCM 1st-byte SoC is 0x11 OR 0x21 when SoC is 20% so we need to add 0x11.
+  SoC_Bytes[1] = evalSoC;
+}
+
+//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+void BATTSCI_evaluateTemperatureByte(uint16_t evalSoC) {
+  /**
+    BATTSCI_evaluateTemperatureByte determines the IMA battery temperature we are sending to the MCM.
+
+    @param      evalSoC     Integer value of SoC in 0.1% increments beginning at 20%. 0 = 20.0%
+    @return                 This function does not return anything.
+  */
+
+  // battTempBATTSCI will be displayed if it won't affect MCM behaviour, as determined by SoC
+  // 0x26 = 08 Deg C
+  // 0x30 = 18 Deg C
+  // 0x31 = 19 Deg C
+  int8_t battTempBATTSCI = temperature_battery_getLatest() + 30;                      //T_MCM = T_actual + 30
+
+  if (evalSoC >= 200)             // 200 = 40% SoC or higher
+  {
+	if (battTempBATTSCI > 0x33) {
+		temperature_Byte = battTempBATTSCI;
+	} else temperature_Byte = 0x33;      // Set temperature to +21 deg C to allow max Assist in 2nd and 3rd
+  } else {
+	if (battTempBATTSCI < 0x30) {
+		temperature_Byte = battTempBATTSCI;
+	} else temperature_Byte = 0x30;      // Set temperature to +18 deg C to reduce max Assist in 2nd and 3rd
+  }
+}
+
+//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+void BATTSCI_calculateSoC()
+{
+  /**
+    BATTSCI_calculateSoC calculates what SoC we want to send the MCM, in order to govern the MCM's behaviour.
+    The MCM changes how the car uses assist and regen depending on the SoC number we send to it.
+
+    @return                 This function does not return anything.
+  */
+  LiBCM_SoC = SoC_getBatteryStateNow_percent();
```

Comment:
Renamed calculatedSoC and oldCalculatedSoC to SoC_sendToMCM and lastSoC_sentToMCM respectively in order to clear up confusion about what those variables represented.

LiBCM_SoC is not sent to MCM, and is not modified -- it is set to SoC_getBatteryStateNow_percent() and never changes, except of course when the SoC_getBatteryStateNow_percent() return value changes.

The only reason I use LiBCM_SoC instead of only using SoC_getBatteryStateNow_percent() was I didn't want to call SoC_getBatteryStateNow_percent() a total of 5 times per frame instead of only 1 time per frame.

If there is no processor overhead expended if that function is called 5 times instead of once then we can entirely delete LiBCM_SoC as a variable.

Reply:
Good point.  In my latest code there is no appreciable processor overhead by calling SoC_getBatteryStateNow_percent() 5 times; this function now immediately returns the latest calculated value.  Yes, please remove your variable.